src/hll.c

```diff
@@ -92,6 +118,182 @@ enum {
+
+	scandesc = systable_beginscan(rel, ExtensionOidIndexId, true,
+								  NULL, 1, entry);
+
+	tuple = systable_getnext(scandesc);
+
+	/* We assume that there can be at most one matching tuple */
+	if (HeapTupleIsValid(tuple))
+		result = ((Form_pg_extension) GETSTRUCT(tuple))->extnamespace;
+	else
+		result = InvalidOid;
+
+	systable_endscan(scandesc);
+
+	heap_close(rel, AccessShareLock);
+
+	return result;
+	/* *INDENT-ON* */
+}
+
+/*
+ * FunctionOid looks for a function that has the given name and the given number
+ * of arguments, and returns the corresponding function's oid.
+ */
+static Oid
+FunctionOid(const char *schemaName, const char *functionName, int argumentCount)
+{
+	FuncCandidateList functionList = NULL;
+	Oid functionOid = InvalidOid;
+
+	char *qualifiedFunctionName = quote_qualified_identifier(schemaName, functionName);
+	List *qualifiedFunctionNameList = stringToQualifiedNameList(qualifiedFunctionName);
+	List *argumentList = NIL;
+	const bool findVariadics = false;
+	const bool findDefaults = false;
+	const bool missingOK = true;
+
+	functionList = FuncnameGetCandidates(qualifiedFunctionNameList, argumentCount,
+										 argumentList, findVariadics,
+										 findDefaults, missingOK);
+
+	if (functionList == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_FUNCTION),
+						errmsg("function \"%s\" does not exist", functionName)));
+	}
+	else if (functionList->next != NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_AMBIGUOUS_FUNCTION),
+						errmsg("more than one function named \"%s\"", functionName)));
+	}
+
+	/* get function oid from function list's head */
+	functionOid = functionList->oid;
+	return functionOid;
+}
+
+/*
+ * MaximizeCostOfHashAggregateWithHLL maximizes the cost of the path if it tries
+ * to run hll aggregate function with hash aggregate.
+ */
+static void
+MaximizeCostOfHashAggregateWithHLL(Path *path)
+{
+	List *varList = pull_var_clause((Node*) path->pathtarget->exprs,
+									PVC_INCLUDE_AGGREGATES);
+	ListCell *varCell = NULL;
+
+	foreach(varCell, varList)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (nodeTag(var) == T_Aggref)
+		{
+			Aggref *aggref = (Aggref *) var;
+
+			// That part will be cached
+			int argCount = list_length(aggref->args);
+			Oid hllId = get_extension_oid(HLL_EXTENSION_NAME, false);
+			Oid hllSchemaOid = get_extension_schema(hllId);
+			const char *hllSchemaName = get_namespace_name(hllSchemaOid);
+			Oid addFunctionId = FunctionOid(hllSchemaName, HLL_ADD_AGG_NAME, argCount);
+			Oid unionFunctionId = FunctionOid(hllSchemaName, HLL_UNION_AGG_NAME, 1);
+
+			if(aggref->aggfnoid == addFunctionId || aggref->aggfnoid == unionFunctionId)
+			{
+				path->total_cost = INT_MAX;
+			}
+		}
+	}
+}
+
+/*
+ * multi_aggregation_restriction_hook is assigned to create_upper_paths_hook to
+ * check whether there exist a path with hash aggregate. If that aggregate is
+ * introduced by hll, it's cost is maximized to force planner to not to select
+ * hash aggregate.
+ */
```

Comment:
we should decide if we need to call previous hook stored int `previous_upper_path_hook` and call it when needed. Otherwise we would block other extensions that register to this hook.